MovingBlocks/Terasology
# i18n: translatable compound messages

This allows translatable strings to contain format arguments. It enables compound messages:

translation resources:
> "tree-punch": "You punch the {0} tree.",
> "tree-species-fir": "fir",
> "tree-species-palm": "palm",

application code:
```java
translate("${#tree-punch}", tree.getSpecies())
```

result:
> _You punch the palm tree._

(This would be a much more compelling demonstration if I knew how to talk about trees in more than one language.)

For this it uses the [MessageFormat class](https://docs.oracle.com/javase/tutorial/i18n/format/messageFormat.html) from the Java library. That also lets it support things like locale-aware date and number formatting. (and probably pluralization? but that interface looks messy and I haven't tested it.)

### How to test

The game-loading error message from #4004 has a compound message. If you know any other languages, you could provide their `.lang` files with a translation for `load-game-error`.

Oh, and check to see that the _existing_ text isn't broken. That's probably the more important thing.


### Outstanding before merging

- [ ] Does this break API compatibility? It adds varargs to several interfaces marked as public `@API`. The changes are _source_ compatible, because it's fine to pass zero varargs, but are they binary compatible?
- [ ] Will need developer documentation. We have a mishmash of the standard Java interface and homegrown formats.
